File: leetcode/medium/96-unique-bst/dp.py
"""
  @author Vortexx2
  Problem 96 - Unique BST

  Runtime - 28 ms
  Memory Usage - 14 MB
"""


class Solution:
  def numTrees(self, n: int) -> int:
    dp = [1] * (n + 1)

    for i in range(2, n + 1):
      dp[i] = 0
      for j in range(i):
        dp[i] += dp[j] * dp[i - j - 1]

    return dp[n]

File: leetcode/medium/96-unique-bst/dp.cpp
/**
 * @file dp.cpp
 * @author Vortexx2
 * @brief Problem 96 - Unique BSTs
 * @date 16-08-2021
 * 
 * Runtime - 0 ms
 * Memory Usage - 6.1 MB
 */
#include <iostream>
#include <vector>

using namespace std;

class Solution
{
public:
  int numTrees(int n)
  {
    vector<int> dp(n + 1, 1);

    for (int i = 2; i <= n; i++)
    {
      dp[i] = 0;
      for (int j = 0; j < i; j++)
        dp[i] += dp[j] * dp[i - j - 1];
    }

    return dp[n];
  }
};